### ISSUE TITLE
[epic] Performance optimization ideas

### ISSUE THREAD

**Issue by borNfreee** [infection/infection#75]
Nikic [has recently published](https://github.com/nikic/PHP-Parser/blob/73be076/doc/component/Performance.markdown) a small article about possible performance optimizations for projects that use PHP-Parser.

I would like to post them here, discuss and step by step try all of them for Infection.

- [x] [Object reuse](https://github.com/nikic/PHP-Parser/blob/73be076/doc/component/Performance.markdown#object-reuse) - [implemented](https://github.com/infection/infection/pull/76)
- [x] [Garbage Collection](https://github.com/nikic/PHP-Parser/blob/73be076/doc/component/Performance.markdown#garbage-collection) - [didn't give](https://github.com/infection/infection/issues/75#issuecomment-346150671) any performance boost. 
- [x] Reuse created mutants to avoid traversing/prettyprinting - [implemented](https://github.com/infection/infection/pull/184)
- [x]  Partially [disabling XDebug](https://github.com/nikic/PHP-Parser/blob/73be076/doc/component/Performance.markdown#xdebug) - [implemented](https://github.com/infection/infection/pull/85). We can disable xDebug for each Symfony Process when e.g. PHPUnit is run against mutated code. Actually, we can disable it even for the main Infection process, and only enable when the initial test suite is executed to generate Code Coverage (this is also a separate Symfony Process in the code).
- [ ] Partially disable `phpdbg`

**Comment by sidz** [comment]
@borNfreee I can pick up "disabling XDebug". If you already have any thoughts feel free to write to me.

**Comment by borNfreee** [comment]
@sidz great, thank you!

**Comment by theofidry** [comment]
tbh I wouldn't touch the garbage collection: it's trivial for the user to disable it if he can afford it by doing `php -d zend.gc_enable=0 vendor/bin/infection`

**Comment by borNfreee** [comment]
I've tested disabling Garbage Collector with different (not big though) projects and for all of them it does not give any performance boost. Sometimes it's even worse with `gc_disable()` - more memory consumption and more execution time.

Marking this point as "won't implement" for now.

**Comment by sidz** [comment]
@borNfreee here is my PR for "Partially disabling XDebug" enhancement: https://github.com/infection/infection/pull/85

suggestions are welcome

**Comment by borNfreee** [comment]
Great! I've also added a point about `phpdbg` to the initial comment to track it

**Comment by borNfreee** [comment]
Didn't do much investigation yet, but want to save another idea to improve performance: reuse created mutants and do not traverse through files again if mutant is already created and original file  wasn't modified between infection runs.

This will allow to avoid traversing and pretty printing:

https://github.com/infection/infection/blob/f2e493795966abf6bce29869dfb478bbf4286447/src/Mutant/MutantCreator.php#L67-L70

@keradus, you had an idea of extracting cache layer from PHPCsFixer, could you please expand on this? Is it reusable? where I can start investigating in PHPCsFixer codebase?

**Comment by keradus** [comment]
Currently, PHP CS Fixer has the cache layer that combines 3 info:
- env (is global for cache); like PHP CS Fixer cares if it was run on 5.x or 7.x
- config (is global for cache); like PHP CS Fixer cares if user want to have enable fixers ABC or XYZ
- files (is not global for cache): checksum for content of original file

Global aspect means that if env or config changed, we need to rerun tool for all files.
If both env and config is the same, we need to rerun tool only for modified files.

Not sure if env part is needed for infection, if not, it could be plugged-off.
Config will be needed, yet instead of fixers you will have mutators.
Files will be used as-is.

One thing to be aware is that if config was modified in any way, eg one fixer was added or removed, we do rerun all fixers, as adding sth means that maybe other fixers would have some hits on "changed" code, as we do actually run all fixers consecutive on output of previous fixer, while you run all mutations on original file only, thus it could be somehow improved to run infection only on added mutators when cache for some of them is already there. Maybe not a must-have for first iteration.

After saying that, be aware that Cache component is still part of main repo. It could be extracted of course, but I would do it only after alignment that Infection is indeed willing and able to take advantage of it (and for dev POC, one could include whole PHP CS Fixer and use cache only).

So, that's briefly how it looks. If you want to discuss in details, let us catch up on gitter, I'm eager to verify possibility and help.

**Comment by BackEndTea** [comment]
>Partially disable phpdbg

Maybe a simple solution would be to advise users using phpdbg to create coverage with phpdbg, and run infection after that without it.

**Comment by borNfreee** [comment]
Could you guys remind why can't we disable `phpdbg` as we do it with `Xdebug`?

If this is not possible, I would close this epic.

**Comment by borNfreee** [comment]
any other performance improvements will be covered by separate issue/PR, closing
